gapic/templates/%namespace/%name_%version/%sub/services/%service/client.py.j2

```diff
@@ -316,36 +322,94 @@ class {{ service.client_name }}(metaclass={{ service.client_name }}Meta):
         """Return the API endpoint used by the client.
 
         Args:
-            api_override (str): The API endpoint override. If specified, this is always the return value of this function.
+            api_override (str): The API endpoint override. If specified, this is always the return value of this function
+            and the other arguments are not used.
             client_cert_source (bytes): The client certificate source used by the client.
+            universe_domain (str): The universe domain used by the client.
             use_mtls_endpoint (str): How to use the MTLS endpoint, which depends also on the other parameters.
                 Possible values are "always", "auto", or "never".
 
         Returns:
             str: The API endpoint to be used by the client.
         """
-
         if api_override is not None:
             api_endpoint = api_override
         elif use_mtls_endpoint == "always" or (use_mtls_endpoint == "auto" and client_cert_source):
+            if universe_domain != {{ service.client_name }}._DEFAULT_UNIVERSE:
+                raise MutualTLSChannelError("MTLS is not supported in any universe other than googleapis.com.")
             api_endpoint = {{ service.client_name }}.DEFAULT_MTLS_ENDPOINT
         else:
-            api_endpoint = {{ service.client_name }}.DEFAULT_ENDPOINT
+            api_endpoint = {{ service.client_name }}.DEFAULT_ENDPOINT_TEMPLATE.format(UNIVERSE_DOMAIN=universe_domain)
         return api_endpoint
+    
+    def _get_universe_domain(client_universe_domain, universe_domain_env):
+        """Return the universe domain used by the client.
+
+        Args:
+            client_universe_domain (str): The universe domain configured via the client options.
+            universe_domain_env (str): The universe domain configured via the "GOOGLE_CLOUD_UNIVERSE_DOMAIN" environment variable.
+
+        Returns:
+            str: The universe domain to be used by the client.
+        """
+        universe_domain = {{ service.client_name }}._DEFAULT_UNIVERSE
+        if client_universe_domain is not None:
+            universe_domain = client_universe_domain
+        elif universe_domain_env is not None:
+            universe_domain = universe_domain_env
+        if len(universe_domain.strip()) == 0:
+            raise ValueError("Universe Domain cannot be an empty string.") 
+        return universe_domain
+
+    @staticmethod
+    def _compare_universes(client_universe, credentials_universe):
+        """Returns True if the universe domain used by the client
+            is the same as the universe domain in credentials.
+
+        Args:
+            client_universe (str): The universe domain configured via the client options.
+            credentials_universe (str): The universe domain in the credentials.
+
+        Returns:
+            bool: True if valid universe, else raise a "ValueError" exception.
```

Comment:
```suggestion
        Returns:
            bool: Returns True if valid universe domain is valid, otherwise False.
        Raises:
            ValueError: If universe domain is an empty string.
```

Reply:
addressed.